nilearn/input_data/nifti_spheres_masker.py

```diff
@@ -85,6 +86,78 @@ def _apply_mask_and_get_affinity(seeds, niimg, radius, allow_overlap,
     return X, A
 
 
+def _get_spheres_rows(seeds, radius, allow_overlap, mask_img):
+    '''Utility function to get only the rows which are occupied by sphere at
+    given seed locations and the provided radius. Rows are in target_affine and 
+    target_shape space.
+
+    Parameters
+    ----------
+    seeds_: List of triplets of coordinates in native space
+        Seed definitions. List of coordinates of the seeds in the same space
+        as target_affine.
+    target_affine: np.ndarray of shape (4,4)
+        Affine for the spheres.
+    target_shape: np.array or list
+        Target shape into which the spheres will be projected. 
+    radius: float
+        Indicates, in millimeters, the radius for the sphere around the seed.
+    allow_overlap: boolean
+        If False, an error is raised if the maps overlaps (ie at least two
+        maps have a non-zero value for the same voxel).
+
+    Returns
+    -------
+    spheres_map: list of lists 
+        The spheres' location (flattened array), one list for each sphere.
+    '''
+    seeds = list(seeds)
+
+    # Compute world coordinates of the mask and of all coordinates
+    mask = _safe_get_data(mask_img)
+    
+    mask_coords = list(zip(*np.where(mask == 0)))
+    mask_coords = np.asarray(list(zip(*mask_coords)))
+    if not np.all(mask == 1):
+        mask_coords = coord_transform(mask_coords[0], mask_coords[1],
+                                    mask_coords[2], mask_img.affine)
+    mask_coords = np.asarray(mask_coords).astype(int).T
+
+    # Calculate brain coordinates
+    brain_coords = list(np.ndindex(*mask_img.shape))
+    brain_coords = np.asarray(list(zip(*brain_coords)))
+    brain_coords = coord_transform(brain_coords[0], brain_coords[1],
+                                  brain_coords[2], mask_img.affine)
+    brain_coords = np.asarray(brain_coords).T
+
+    clf = neighbors.NearestNeighbors(radius=radius)
+    adjacency = clf.fit(brain_coords).radius_neighbors_graph(seeds)
+    adjacency = adjacency.tolil()
+
+    # Include the voxel containing the seed itself
+    brain_coords = brain_coords.astype(int).tolist()
+    # Get the mask indices (in terms of brain coords)
+    mask_idx = [brain_coords.index(i) for i in mask_coords.tolist()]
```

Comment:
this loop looks a bit terrible computationally. Wouldn't it be more efficient to store mask as a binary array and use numpy indexing ?